src/client.rs

```diff
@@ -23,83 +32,88 @@ pub struct CycleTimes {
 /// // TODO: make example
 /// ```
 #[derive(Debug)]
-pub struct Client<T: JackHandler> {
+pub struct Client {
+    client: *mut j::jack_client_t,
+}
+
+#[derive(Debug)]
+pub struct ActiveClient<JH: JackHandler> {
     client: *mut j::jack_client_t,
-    handler: *mut T,
-    status: ClientStatus,
+    handler: *mut (JH, *mut j::jack_client_t),
+}
+
+unsafe impl JackClient for Client {
+    unsafe fn client_ptr(&self) -> *mut j::jack_client_t { self.client }
+}
+
+unsafe impl<JH: JackHandler> JackClient for ActiveClient<JH> {
+    unsafe fn client_ptr(&self) -> *mut j::jack_client_t { self.client }
 }
 
-impl<T: JackHandler> Client<T> {
-    /// The maximum length of the Jack client name string. Unlike the "C" Jack
-    /// API, this does not take into account the final `NULL` character and
-    /// instead corresponds directly to `.len()`. This value is constant.
-    pub fn name_size() -> usize {
-        let s = unsafe { j::jack_client_name_size() - 1 };
-        s as usize
+pub unsafe trait JackClient: Sized {
+
+    #[inline(always)]
+    unsafe fn client_ptr(&self) -> *mut j::jack_client_t;
+
+    /// Manually close the client, deactivating if necessary.
+    /// This will happen automatically on drop.
+    fn close(self) -> () { drop(self) }
+
+    /// The sample rate of the jack system, as set by the user when jackd was
+    /// started.
+    fn sample_rate(&self) -> usize {
+        let srate = unsafe { j::jack_get_sample_rate(self.client_ptr()) };
+        srate as usize
     }
 
-    /// The buffer size of a port type
-    ///
-    /// # Unsafe
+    /// The current maximum size that will every be passed to the process
+    /// callback.
     ///
-    /// * This function may only be called in a buffer size callback.
```

Comment:
I'm unsure about what you meant by "may only be called in a buffer size callback", and I'm fairly certain I haven't implemented this. However, I'm not sure that this function is strictly necessary now that we have `ProcessScope` providing the number of frames. It might still be useful for clients who want to pre-allocate for out-of-place buffer transformations.